### PR TITLE
[QC][Ascend] Add INT8 W8A16 RMSNorm operator

### PR DESCRIPTION
**Pull request by July-h5kf3**
## Summary
Adds an Ascend-only W8A16 RMSNorm operator (`rms_norm_w8a16_int8`). Activation is FP16/BF16; weight is grouped INT8 plus per-group scale (`group_size=128`) because Ascend 910 UB / DataCopy cannot load FP8.

Dispatch:
- Power-of-two `N <= 4096`: 1D kernel with `BLOCK_M` rows so INT8 weight + scale are loaded once per program and reused. Mid-size M uses `BLOCK_M=8`; small and very large M use `BLOCK_M=2` (higher occupancy).
- Power-of-two `N <= 8192`: 1D row kernel. Full-row `BLOCK_M` overflows Ascend UB. Scale is loaded uniquely and broadcast via reshape (no `cols // group_size` gather; no `(G, 128)` FP32 2D tile).
- Otherwise: tiled 1D kernel (`GROUPS_PER_TILE=64`) to stay in UB.

This replaces closed #5709 on a new `Ascend/` branch.

## Testing
- Parametrized tests over `(1, 4096)`, `(128, 4096)`, `(512, 4096)`, `(64, 8192)`, `(1, 16384)`, `(1, 32768)` in bf16 (6 cases)
- Validated against dequantized INT8 weight + `torch.nn.functional.rms_norm` on device
- Tested on: Ascend 910B4

## Performance
Accuracy command: `pytest tests/test_rms_norm.py -k test_rms_norm_w8a16_int8`

Performance is **NPUGraph** (`torch.npu.NPUGraph` capture + replay, median of 8 replays) on Ascend 910B4 (64GB HBM), CANN, torch_npu. FlagGems `--mode cudagraph` is CUDA Graph and does not run on this NPU.

Bench shapes: `(1, 4096)`, `(16, 4096)`, `(64, 4096)`, `(256, 4096)`, `(1024, 4096)`, `(1, 8192)`, `(64, 8192)`, `(256, 8192)`, `(1, 16384)`, `(64, 16384)`. Speedup > 1 means W8A16 is faster.

**bfloat16 vs torch `F.rms_norm`:**

| Shape | Torch (ms) | Gems W8A16 (ms) | Speedup |
| --- | --- | --- | --- |
| 1x4096 | 0.0202 | 0.0041 | 5.00x |
| 16x4096 | 0.0415 | 0.0042 | 9.85x |
| 64x4096 | 0.0580 | 0.0053 | 10.92x |
| 256x4096 | 0.0716 | 0.0104 | 6.89x |
| 1024x4096 | 0.1134 | 0.0242 | 4.68x |
| 1x8192 | 0.0210 | 0.0037 | 5.63x |
| 64x8192 | 0.0657 | 0.0083 | 7.95x |
| 256x8192 | 0.0890 | 0.0179 | 4.96x |
| 1x16384 | 0.0266 | 0.0053 | 5.01x |
| 64x16384 | 0.0644 | 0.0121 | 5.34x |
| **Arithmetic Mean** | — | — | **6.62x** |

**bfloat16 vs Gems `flag_gems.rms_norm`:**

| Shape | Gems BF16 (ms) | Gems W8A16 (ms) | Speedup |
| --- | --- | --- | --- |
| 1x4096 | 0.0040 | 0.0041 | 0.98x |
| 16x4096 | 0.0048 | 0.0042 | 1.13x |
| 64x4096 | 0.0089 | 0.0053 | 1.67x |
| 256x4096 | 0.0170 | 0.0104 | 1.63x |
| 1024x4096 | 0.0480 | 0.0242 | 1.98x |
| 1x8192 | 0.0049 | 0.0037 | 1.30x |
| 64x8192 | 0.0094 | 0.0083 | 1.14x |
| 256x8192 | 0.0212 | 0.0179 | 1.18x |
| 1x16384 | 0.0072 | 0.0053 | 1.35x |
| 64x16384 | 0.0189 | 0.0121 | 1.56x |
| **Arithmetic Mean** | — | — | **1.39x** |

**Overall Mean Speedup: 6.62x vs torch (10 cases); 1.39x vs Gems BF16 (10 cases)**

## Multi-backend Testing

| Backend | Accuracy | Benchmark | Speedup | Cases | Notes |
| --- | --- | --- | --- | --- | --- |
| Nvidia | — | — | — | — | Not tested (Ascend-only INT8 path; NVIDIA keeps FP8) |
| Tianshu | — | — | — | — | Not tested |
| Muxi | — | — | — | — | Not tested |
| Ascend 910B4 | PASS | PASS | 6.62x vs torch / 1.39x vs Gems BF16 | 6 acc / 10 bench | NPUGraph replay, bf16 |
| Hygon | — | — | — | — | Not tested |

## Files Changed
- `src/flag_gems/runtime/backend/_ascend/ops/rms_norm_w8a16_int8.py`: Ascend W8A16 RMSNorm kernel and dispatch
- `src/flag_gems/runtime/backend/_ascend/ops/__init__.py`: Override `rms_norm_w8a16_int8` in the backend
- `tests/test_rms_norm.py`: Accuracy tests (Ascend INT8 / NVIDIA FP8)
- `benchmark/test_rms_norm.py`: Performance benchmark aligned with the NVIDIA W8A16 path

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: src/flag_gems/runtime/backend/_ascend/ops/rms_norm_w8a16_fp8.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+    dim = x.ndim - len(normalized_shape)
+    M = math.prod(x.shape[:dim])
+    N = math.prod(normalized_shape)
+    if N % group_size != 0:
+        raise ValueError(
+            f"normalized_shape product {N} must be divisible by group_size={group_size}"
+        )
+    if weight_fp8.dtype != torch.int8:
+        raise TypeError(
+            f"Ascend W8A16 RMSNorm expects INT8 weight, got {weight_fp8.dtype}"
```
Comment (0x45f) [inline src/flag_gems/runtime/backend/_ascend/ops/rms_norm_w8a16_fp8.py]: So this API should be `rms_norm_w8a16_int8`?

Comment (July-h5kf3) [reply src/flag_gems/runtime/backend/_ascend/ops/rms_norm_w8a16_fp8.py]: Yes, you are right. This Ascend API should be renamed to `rms_norm_w8a16_int8`. The current CANN/Triton-Ascend data-movement path on Ascend 910 cannot load FP8 weights from GM into UB, so the kernel cannot directly consume FP8 tensors. Therefore, we use grouped INT8 weights with one scale per 128 elements and apply the scale during RMSNorm. I will rename the API accordingly.

Comment (July-h5kf3) [reply src/flag_gems/runtime/backend/_ascend/ops/rms_norm_w8a16_fp8.py]: Yes, you are right. I renamed the Ascend API to `rms_norm_w8a16_int8`. On Ascend 910, the current CANN/Triton-Ascend data-movement path cannot load FP8 weights from GM into UB, so this kernel cannot consume FP8 tensors. It therefore uses grouped INT8 weights with one scale per 128 elements and applies the scale during RMSNorm.

**Review by 0x45f — APPROVED** [review]
